processing/src/main/java/io/druid/segment/filter/InFilter.java

```diff
@@ -85,24 +86,48 @@ public boolean apply(String input)
   @Override
   public ValueMatcher makeMatcher(ValueMatcherFactory factory)
   {
-    return factory.makeValueMatcher(
-        dimension, new Predicate<String>()
-        {
-          @Override
-          public boolean apply(String input)
-          {
-            if (extractionFn != null) {
-              input = extractionFn.apply(input);
-            }
-            return values.contains(Strings.nullToEmpty(input));
-          }
-        }
-    );
+    return factory.makeValueMatcher(dimension, getPredicate());
   }
 
   @Override
   public boolean supportsBitmapIndex(BitmapIndexSelector selector)
   {
     return selector.getBitmapIndex(dimension) != null;
   }
+
+  private DruidCompositePredicate getPredicate()
+  {
+    if (extractionFn == null) {
+      return new DruidCompositePredicate()
+      {
+        @Override
+        public boolean apply(Object inputObj)
+        {
+          String input = inputObj == null ? null : inputObj.toString();
+          return values.contains(Strings.nullToEmpty(input));
+        }
+
+        @Override
+        public boolean applyLong(long value)
+        {
+          return values.contains(String.valueOf(value));
```

Comment:
It looks like it's fastest to use binary search on `long[]` until 16 elements or so, and then use `HashSet<Long>` after that.

```
LongSetBenchmark.array                   1  thrpt   10  222005790.961 ± 22300054.965  ops/s
LongSetBenchmark.array                   2  thrpt   10  221364061.494 ± 15765300.290  ops/s
LongSetBenchmark.array                   4  thrpt   10  154949143.962 ±  7532157.144  ops/s
LongSetBenchmark.array                   8  thrpt   10  147140083.071 ± 10224373.832  ops/s
LongSetBenchmark.array                  16  thrpt   10  121571276.511 ±  7832101.665  ops/s
LongSetBenchmark.array                  32  thrpt   10  102713662.860 ±  7359706.657  ops/s
LongSetBenchmark.array                  64  thrpt   10   93137148.636 ±  7481012.126  ops/s
LongSetBenchmark.array                 128  thrpt   10   84266319.187 ±  4305522.254  ops/s
LongSetBenchmark.array                 256  thrpt   10   72152173.792 ±  2866558.752  ops/s
LongSetBenchmark.array                 512  thrpt   10   71878361.594 ±  6460088.134  ops/s
LongSetBenchmark.array                1024  thrpt   10   66977923.966 ±  4864061.239  ops/s
LongSetBenchmark.arrayList               1  thrpt   10  134397445.490 ± 11339183.195  ops/s
LongSetBenchmark.arrayList               2  thrpt   10  140457560.970 ±  7030480.663  ops/s
LongSetBenchmark.arrayList               4  thrpt   10   86483528.676 ±  5659909.339  ops/s
LongSetBenchmark.arrayList               8  thrpt   10   79038125.180 ±  4540887.373  ops/s
LongSetBenchmark.arrayList              16  thrpt   10   74278843.660 ±  6377259.580  ops/s
LongSetBenchmark.arrayList              32  thrpt   10   60456055.375 ±  3355174.428  ops/s
LongSetBenchmark.arrayList              64  thrpt   10   52944752.462 ±  3333871.836  ops/s
LongSetBenchmark.arrayList             128  thrpt   10   44796549.858 ±  3266666.774  ops/s
LongSetBenchmark.arrayList             256  thrpt   10   40471751.788 ±  2885321.752  ops/s
LongSetBenchmark.arrayList             512  thrpt   10   38497404.764 ±  1504354.437  ops/s
LongSetBenchmark.arrayList            1024  thrpt   10   37686543.130 ±  3134658.908  ops/s
LongSetBenchmark.hashSet                 1  thrpt   10  156736734.488 ±  8651301.827  ops/s
LongSetBenchmark.hashSet                 2  thrpt   10  147066347.299 ± 15020205.076  ops/s
LongSetBenchmark.hashSet                 4  thrpt   10  135561948.620 ± 12590945.334  ops/s
LongSetBenchmark.hashSet                 8  thrpt   10  136032740.412 ± 17352408.843  ops/s
LongSetBenchmark.hashSet                16  thrpt   10  147755956.329 ± 16008970.374  ops/s
LongSetBenchmark.hashSet                32  thrpt   10  146337051.449 ± 12207525.510  ops/s
LongSetBenchmark.hashSet                64  thrpt   10  149583904.399 ± 15493885.057  ops/s
LongSetBenchmark.hashSet               128  thrpt   10  150574341.903 ± 13577283.581  ops/s
LongSetBenchmark.hashSet               256  thrpt   10  143885737.975 ±  8454938.219  ops/s
LongSetBenchmark.hashSet               512  thrpt   10  149395185.940 ±  8307174.413  ops/s
LongSetBenchmark.hashSet              1024  thrpt   10  147016100.707 ± 14865907.098  ops/s
LongSetBenchmark.immutableSet            1  thrpt   10  188235322.562 ± 17044931.922  ops/s
LongSetBenchmark.immutableSet            2  thrpt   10  128273217.344 ± 12222080.845  ops/s
LongSetBenchmark.immutableSet            4  thrpt   10  125857860.943 ±  9402357.117  ops/s
LongSetBenchmark.immutableSet            8  thrpt   10  113090901.134 ±  9958665.897  ops/s
LongSetBenchmark.immutableSet           16  thrpt   10  113484312.472 ±  7660106.185  ops/s
LongSetBenchmark.immutableSet           32  thrpt   10  119513034.435 ± 12465977.793  ops/s
LongSetBenchmark.immutableSet           64  thrpt   10  111235193.583 ±  8644656.497  ops/s
LongSetBenchmark.immutableSet          128  thrpt   10  117649659.156 ± 10472582.443  ops/s
LongSetBenchmark.immutableSet          256  thrpt   10  108513247.457 ±  5690035.363  ops/s
LongSetBenchmark.immutableSet          512  thrpt   10  114346871.358 ±  7200422.403  ops/s
LongSetBenchmark.immutableSet         1024  thrpt   10  116463481.892 ±  7313632.478  ops/s
LongSetBenchmark.treeSet                 1  thrpt   10  135210460.681 ± 14016216.045  ops/s
LongSetBenchmark.treeSet                 2  thrpt   10  119224390.011 ± 11906977.409  ops/s
LongSetBenchmark.treeSet                 4  thrpt   10  117449065.095 ±  9740217.346  ops/s
LongSetBenchmark.treeSet                 8  thrpt   10   90261839.524 ±  5523265.799  ops/s
LongSetBenchmark.treeSet                16  thrpt   10   78669891.810 ±  7033409.741  ops/s
LongSetBenchmark.treeSet                32  thrpt   10   64278431.773 ± 10560825.368  ops/s
LongSetBenchmark.treeSet                64  thrpt   10   59146787.931 ±  4773434.788  ops/s
LongSetBenchmark.treeSet               128  thrpt   10   53214433.112 ±  4658405.269  ops/s
LongSetBenchmark.treeSet               256  thrpt   10   45396537.632 ±  3189184.698  ops/s
LongSetBenchmark.treeSet               512  thrpt   10   42790640.123 ±  3938454.233  ops/s
LongSetBenchmark.treeSet              1024  thrpt   10   38893957.130 ±  2721455.556  ops/s
```